astronet/direct_tensor/predict.py

```diff
@@ -0,0 +1,382 @@
+        Dict of {feature_name: feature_value} for features that are or depend on views of the
+        lightcurve.
+    fold_num: array[float]
+        Array containing the fold number of each point in the lightcurve. Used to calculate the
+        total number of folds in the lightcurve for the "n_folds" feature.
+    """
+    tag = "" if breakspace is None else f"_{breakspace}"
+    all_features = {}
+
+    det_time, det_flux, transit_mask = preprocess.detrend_and_filter(
+        tic, time, flux, period, epoch, duration, breakspace
+    )
+    folded_time, folded_flux, fold_num, normal_transit_mask = (
+        preprocess.phase_fold_and_sort_light_curve(
+            det_time, det_flux, transit_mask, period, epoch
+        )
+    )
+    odd_mask = fold_num % 2 == 1
+    even_mask = fold_num % 2 == 1
+
+    all_features.update(
+        global_features(tic, folded_time, folded_flux, normal_transit_mask, period)
+    )
+
+    loc_features, local_scale, local_depth = local_features(
+        tic, folded_time, folded_flux, period, duration
+    )
+    all_features.update(loc_features)
+
+    for aperture, (ap_time, ap_flux) in aperture_fluxes.items():
+        ap_det_time, ap_det_flux, ap_transit_mask = preprocess.detrend_and_filter(
+            tic, ap_time, ap_flux
+        )
+        ap_folded_time, ap_folded_flux, _, _ = (
+            preprocess.phase_fold_and_sort_light_curve(
+                ap_det_time, ap_det_flux, ap_transit_mask, period, epoch
+            )
+        )
+        all_features.update(
+            aperture_features(
+                aperture,
+                tic,
+                ap_folded_time,
+                ap_folded_flux,
+                period,
+                duration,
+                local_scale,
+                local_depth,
+            )
+        )
+
+    all_features.update(
+        odd_features(
+            odd_mask,
+            tic,
+            folded_time,
+            folded_flux,
+            period,
+            duration,
+            local_scale,
+            local_depth,
+        )
+    )
+    all_features.update(
+        even_features(
+            even_mask,
+            tic,
+            folded_time,
+            folded_flux,
+            period,
+            duration,
+            local_scale,
+            local_depth,
+        )
+    )
+
+    sec_features, secondary_scale = secondary_features(
+        tic, folded_time, folded_flux, period, duration, local_scale, local_depth
+    )
+    all_features.update(sec_features)
+
+    all_features.update(
+        sample_segments_features(
+            tic,
+            folded_time,
+            folded_flux,
+            fold_num,
+            odd_mask,
+            even_mask,
+            period,
+            duration,
+        )
+    )
+
+    double_fold_time, double_fold_flux, _, _ = (
+        preprocess.phase_fold_and_sort_light_curve(
+            det_time, det_flux, transit_mask, period * 2, epoch - period / 2
```

Comment:
Why do we subtract `period / 2` from `epoch`?

Reply:
It's to center the double period view. There are 2 transits in the view, and if we didn't have this subtraction, one of the transits would be centered while the other was split between the very start and very end of the view.

See [this image](https://content.cld.iop.org/journals/1538-3881/165/3/95/revision1/ajacad85f8_hr.jpg) from the 2023 paper on the most recent model.
![Diagram showing architecture of the astronet model.](https://content.cld.iop.org/journals/1538-3881/165/3/95/revision1/ajacad85f8_hr.jpg)